lib/skate/detours/snapshot_serde.ex

```diff
@@ -27,4 +31,280 @@ defmodule Skate.Detours.SnapshotSerde do
   """
   def id_from_snapshot(%{"context" => %{"uuid" => id}}), do: id
   def id_from_snapshot(%{"context" => %{}}), do: nil
+
+  @doc """
+  Builds XState Snapshot from Detours Database object
+  """
+  def serialize(%Detour{} = detour) do
+    validate_serialized_snapshot(
+      %{
+        "value" => state_from_detour(detour),
+        "status" => "active",
+        "context" => context_from_detour(detour),
+        "children" => snapshot_children_from_detour(detour),
+        "historyValue" => %{}
+      },
+      detour
+    )
+  end
+
+  defp validate_serialized_snapshot(
+         serialized_snapshot,
+         %Detour{id: id, state: state}
+       ) do
+    if serialized_snapshot === state do
+      serialized_snapshot
+    else
+      log_fallback(id)
+      Logger.info(MapDiff.diff(state, serialized_snapshot))
+      state
+    end
+  end
+
+  # For each of these retrieve functions, the first function is the one
+  # that will look for the correct field on the db object, once the detour state
+  # is properly distributed across db tables / columns.
+  #
+  # The second function is the fallback that uses the existing state snapshot.
+  #
+  # Note that the first function is only stubbed out and will be significantly updated
+  # when the new fields are added to the db. For example, `get_route_from_db_detour`
+  # suggests that the field `route` will be surfaced at the top level of the detour
+  # but actually, it may be nested under `route_pattern` or however else we organize
+  # it in the db.
+
+  defp context_from_detour(%Detour{} = detour) do
+    %{
+      "uuid" => uuid_from_detour(detour),
+      "route" => route_from_detour(detour),
+      "routePattern" => routepattern_from_detour(detour),
+      "routePatterns" => routepatterns_from_detour(detour),
+      "startPoint" => startpoint_from_detour(detour),
+      "endPoint" => endpoint_from_detour(detour),
+      "waypoints" => waypoints_from_detour(detour),
+      "nearestIntersection" => nearestintersection_from_detour(detour),
+      "detourShape" => detourshape_from_detour(detour),
+      "finishedDetour" => finisheddetour_from_detour(detour),
+      "editedDirections" => editeddirections_from_detour(detour),
+      "selectedDuration" => selectedduration_from_detour(detour),
+      "selectedReason" => selectedreason_from_detour(detour)
+    }
+    |> Enum.filter(fn {_, v} -> v != nil end)
+    |> Enum.into(%{})
```

Comment:
suggestion(non-blocking): Love it! just wondering, what do you think about using [`Map.filter/2`](https://hexdocs.pm/elixir/Map.html#filter/2)?

```suggestion
    |> Map.filter(fn {_, v} -> v != nil end)
```

https://github.com/devonestes/fast-elixir#filtering-maps-code
says `:maps.filter/2` is faster, by up to 1.3x, but `:maps.filter/2` has the fn first, and the map argument second, which wouldn't work well with Elixir's pipes. (I know we're not at the point of chasing down every bit of performance right now, I just use it as a reference)

Reply:
Agh rats, I did look up `Map.filter` vs `Enum.filter |> Enum.into` and interpreted that the latter was more efficient, but can't find my primary source 🫠  But I'm more open to :maps.filter. I did consider that